### PR TITLE
Implementation of the Route53 specific symlink style ALIAS

### PR DESCRIPTION
**Pull request by ross**
Now that got around to playing with Record type registration https://github.com/octodns/octodns/pull/889 I was able to poke around at things and implement support for the Route53 ALIAS type, which is not the same as octoDNS's alias. It's specific to Route53 and is essentially a symlink to something. As-is this will only support the "Alias to another record in this hosted zone". 

In order to support more types we need to know what data Route53/boto returns for them, e.g. 

```python
{'AliasTarget': {'DNSName': 'exxampled.com.',
                 'EvaluateTargetHealth': False,
                 'HostedZoneId': 'Z05116462IZ3MBXPYGJTH'},
 'Name': 'alias.exxampled.com.',
 'Type': 'A'}
```

The config for the new provider-specific type looks something like this, the `type` is a convention and not anything magic/dynamic.

```yaml
alias:
  ttl: 33
  type: Route53Provider/ALIAS
  values:
    - name: 
      type: A
    - name: www
      type: AAAA
```

Interestingly the type has no TTL. It's the only record type I've ever run across that doesn't have one. That makes some sense given its nature of being a symlink it just would use the TTL of whatever is being pointed at. That's a complication for octoDNS records which require one. Anyway, they're completely ignored with a provider `_include_change`.

/cc https://github.com/octodns/octodns/pull/889 which this requries since it needs to register a new Record type, CI will fail until that's merged
/cc https://github.com/octodns/octodns-route53/issues/24 which requested ALIAS support

### CODE REVIEW

**Comment by slandry90** [comment]
I think it would be worth adding 'HostedZoneId' as an option to the value, the most common use case for the alias record is to point to an elb or nlb cname, both of these cases use a known hosted zone id value https://docs.aws.amazon.com/general/latest/gr/elb.html. My concern here would be how to handle provider to provider syncing where the alias record wouldn't really be modifiable.

**Comment by ross** [comment]
> I think it would be worth adding 'HostedZoneId' as an option to the value, the most common use case for the alias record is to point to an elb or nlb cname, both of these cases use a known hosted zone id value

Yeah, there's a comment specifically about the need to support the other alias types: https://github.com/octodns/octodns-route53/pull/25/files#diff-babf262ee2f3b2410ea3b88fbe7fbad1da9a7bf580276e594910bfa6d373b0cfR1039-R1041. I hadn't yet dug into what the data looks like for them so didn't know what needed to happen. 

> My concern here would be how to handle provider to provider syncing where the alias record wouldn't really be modifiable.

Yeah other providers will completely ignore the records as they're not a type they support, nor would it be possible for them to since it has internal AWS details for things like ELBs.

**Comment by slandry90** [comment]
> Yeah, there's a comment specifically about the need to support the other alias types: https://github.com/octodns/octodns-route53/pull/25/files#diff-babf262ee2f3b2410ea3b88fbe7fbad1da9a7bf580276e594910bfa6d373b0cfR1039-R1041. I hadn't yet dug into what the data looks like for them so didn't know what needed to happen.

The data format is the same for any Alias in route53. The difference is only in what hosted zone id is passed. There's a handful of different use cases for what can be passed for hosted zone id, the route53 backend takes the id and does a lookup for the dnsname that is passed and returns an `InvalidChangeBatch` if the dnsname isn't found in the hosted zone.

Let me know if you need any specific use cases for this and I can help supply

**Comment by slandry90** [comment]
Thanks @ross! I can run some testing on this later today and report back

**Comment by slandry90** [comment]
Tested this in our environment and everything looks really good. Left one comment on supporting additional aws resources and also noticed that the "evaluate_target_health" param must be a boolean. Otherwise this looks great and covers the r53 alias use case very well. Thanks!

**Comment by ross** [comment]
> Left one comment on supporting additional aws resources and also noticed that the "evaluate_target_health" param must be a boolean

Don't think this comment came through. Maybe an un-submitted review?

File: octodns_route53/__init__.py [hunk 1]
```diff
@@ -241,6 +269,49 @@ def _values_for_SRV(self, record):
...
+
+    def __init__(self, provider, hosted_zone_id, record, value, creating):
+        super().__init__(provider, record, creating)
+        self.hosted_zone_id = hosted_zone_id
+        self.fqdn = record.fqdn
+        name = value.name
+        if name:
+            if 'amazonaws.com.' in name:
+                # It's a service symlink, just use it as is
+                self.target_name = name
```
Comment (slandry90) [inline octodns_route53/__init__.py:282]: This restricts usage of the ALIAS to alb/elb/nlb/s3/vpc. There are also two other cases that I know of, cloudfront - (cloudfront.net.), and elasticbeanstalk - (elasticbeanstalk.com.)

There may be additional use cases for global accelerator and api gateway but I don't have those use cases but I'm struggling to find any good documentation on the alias endpoint patterns on those.

https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-values-alias-common.html#rrsets-values-alias-common-target

Comment (ross) [reply octodns_route53/__init__.py:282]: > This restricts usage of the ALIAS to alb/elb/nlb/s3/vpc. There are also two other cases that I know of, cloudfront - (cloudfront.net.), and elasticbeanstalk - (elasticbeanstalk.com.)

Thanks. I wondered if there were others ...

> There may be additional use cases for global accelerator and api gateway but I don't have those use cases but I'm struggling to find any good documentation on the alias endpoint patterns on those.

but had no luck finding a list of the possibilities or a better way to differentiate that was more reliable. 

Will look into broader support.

File: octodns_route53/record.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    @classmethod
+    def validate(cls, data, _type):
+        if not isinstance(data, (list, tuple)):
+            data = (data,)
+        reasons = []
+        for value in data:
+            if 'type' not in value:
+                reasons.append('missing type')
+            if 'amazonaws.com.' in (value.get('name') or ''):
```
Comment (slandry90) [inline octodns_route53/record.py:15]: @ross think you missed updating the service_fqdn list here

Comment (ross) [reply octodns_route53/record.py:15]: Good catch! I'd totally forgotten about the record class. I've reshuffled things around to move the check over into the class and then use it in all 3 places. Added some tests that would have caught the problem as well.

Comment (slandry90) [reply octodns_route53/record.py]: Thanks! Can confirm everything looks good in live testing as well!

**Comment by slandry90** [comment]
@ross what's left to get this merged up?

**Comment by ross** [comment]
> @ross what's left to get this merged up?

Bascially review. It was a meaty enough change that I am looking for someone on @octodns/review to 👁️ it over before :shipit:

**Review by ross — COMMENTED** [review]
Going to timeout on this and move forward.

**Comment by viranch** [comment]
Apologies I tried gleaning at the implementation but couldn't really make out what's going on due to lack of r53 familiarity.

**Pull request merged**